File: Packages of the code extras/po-community-mcp-main/python/tools/patient_id_tool.py
from typing import Annotated

from mcp.server.fastmcp import Context
from pydantic import Field

from fhir_client import FhirClient
from fhir_utilities import get_fhir_context
from mcp_utilities import create_text_response


async def find_patient_id(
    firstName: Annotated[str, Field(description="The patient's first name")],  # noqa: N803
    lastName: Annotated[str | None, Field(description="The patient's last name. This is optional")] = None,  # noqa: N803
    ctx: Context = None,
) -> str:
    patients = await _patient_searcher(ctx, firstName, lastName)
    if not patients:
        patients = await _patient_searcher(ctx, lastName, firstName)

    if patients and len(patients) > 1:
        return create_text_response("More than one patient was found. Provide more details.", is_error=True)

    if patients and patients[0].get("id"):
        return create_text_response(patients[0]["id"])

    return create_text_response("No patient could be found with that name", is_error=True)


async def _find_patient(
    ctx: Context,
    search_first_name: str | None,
    search_last_name: str | None,
) -> list[dict] | None:
    fhir_context = get_fhir_context(ctx)
    if not fhir_context:
        raise ValueError("The fhir context could not be retrieved")

    fhir_client = FhirClient(base_url=fhir_context.url, token=fhir_context.token)

    search_parameters: dict[str, str] = {}
    if search_first_name:
        search_parameters["given"] = search_first_name
    if search_last_name:
        search_parameters["family"] = search_last_name

    bundle = await fhir_client.search("Patient", search_parameters)
    if not bundle or not bundle.get("entry"):
        return None

    return [
        entry["resource"]
        for entry in bundle["entry"]
        if entry.get("resource")
    ]

File: Packages of the code extras/po-community-mcp-main/dotnet/Po.Community.Core/McpTools/PatientIdTool.cs
using Hl7.Fhir.Model;
using Hl7.Fhir.Rest;
using Microsoft.AspNetCore.Http;
using ModelContextProtocol.Protocol;
using ModelContextProtocol.Server;
using Po.Community.Core.Extensions;
using Po.Community.Core.Models;
using Po.Community.Core.Utilities;

namespace Po.Community.Core.McpTools;

public class PatientIdTool : IMcpTool
{
    private const string FirstNameParameter = "firstName";
    private const string LastNameParameter = "lastName";

    public string Name { get; } = "FindPatientId";

    public string? Description { get; } = "Finds a patient id given a first name and last name";

    public List<McpToolArgument> Arguments { get; } =
    [
        new McpToolArgument
        {
            Type = "string",
            Name = FirstNameParameter,
            Description = "The patient's first name",
            IsRequired = true,
        },
        new McpToolArgument
        {
            Type = "string",
            Name = LastNameParameter,
            Description = "The patient's last name. This is optional",
            IsRequired = false,
        },
    ];

    public async Task<CallToolResult> HandleAsync(
        HttpContext httpContext,
        McpServer mcpServer,
        IServiceProvider serviceProvider,
        CallToolRequestParams context
    )
    {
        var firstName = context.GetRequiredArgumentValue(FirstNameParameter);
        var lastName = context.GetArgumentValueOrNull(LastNameParameter);

        var fhirClient = httpContext.CreateFhirClientWithContext();

        var (firstError, firstPatient) = await SearchPatientAsync(fhirClient, firstName, lastName);
        if (firstError is not null)
        {
            return firstError;
        }

        if (firstPatient is not null)
        {
            return McpToolUtilities.CreateTextToolResponse(firstPatient.Id.GetOrThrowIfNull());
        }

        var (secondError, secondPatient) = await SearchPatientAsync(
            fhirClient,
            lastName,
            firstName
        );

        if (secondError is not null)
        {
            return secondError;
        }

        return secondPatient is not null
            ? McpToolUtilities.CreateTextToolResponse(secondPatient.Id.GetOrThrowIfNull())
            : McpToolUtilities.CreateTextToolResponse(
                "No patient could be found with that name",
                isError: true
            );
    }

    private async Task<(CallToolResult? errorResponse, Patient? patient)> SearchPatientAsync(
        FhirClient fhirClient,
        string? firstName,
        string? lastName
    )
    {
        var parameters = new List<string>();
        if (!string.IsNullOrWhiteSpace(firstName))
        {
            parameters.Add($"given={firstName}");
        }

        if (!string.IsNullOrWhiteSpace(lastName))
        {
            parameters.Add($"family={lastName}");
        }

        var response = await fhirClient.SearchAsync<Patient>([.. parameters]);
        if (
            response?.Entry is not { Count: > 0 }
            || response.Entry[0].Resource is not Patient patient
        )
        {
            return (null, null);
        }

        if (response.Entry.Count > 1)
        {
            return (
                McpToolUtilities.CreateTextToolResponse(
                    "More than one patient found. Provide additional details.",
                    isError: true
                ),
                null
            );
        }

        return (null, patient);
    }
}
